Enhance locks

evanchaoli opened the pull request:
## Changes proposed by this PR

Concourse highly relies on db locks to coordinate works across ATCs. But all locks shares a single db connection and a single mutex.

For example, when GC component is acquiring a BatchLock, the other go-routing that is trying to acquire a VolumeCreating lock has to wait the global mutex.

The global mutex is a performance bottleneck on a busy deployment. For example, on our prod cluster, step initialization were always slow, and I noticed that the more resource checking locks hold, then less volume creating locks could be acquired. Thus I created a private build that give a volume-create lock a separate mutex. The private build has been significantly reduced step initialization on our prod cluster.

This PR will give certain lock types dedicate mutex, so that one lock type won't impact others.

During my load tests, I also found that, after giving each lock type a dedicate mutex, locks could be quickly acquired, but one ATC could get extremely busy while other ATCs were idle. That was because some component might be always run on a single ATC. For example, if Lidar scanner is running on ATC1, then ATC1 needs to acquire a lot of resource checking locks, then ATC1 has less chance to acquire `BatchLock("lidar")` again because of the global mutex, so that other ATC will be more likely to run Lidar scanner in next round.

So we need to enhancement component as well to ensure each ATC to have the same chance to run a component.

* [x] Enhance locks

## Notes to reviewer

<!--
If needed, leave any special pointers for reviewing or testing your PR.
-->

## Release Note

<!--
Your PR title will be directly included in the release notes when it ships in
the next Concourse release. It should briefly describe the PR in [imperative
mood]. Please refrain from adding prefixes like 'feature:', and don't include a
period at the end.

Within this section you may supply a list of extra information to include in
the release notes in addition to the pull request title.

Example title: Introduce new pipeline UI algorithm

Example notes:

* Reticulating splines is the new process Concourse uses to create the network
  of lines between jobs.
* Combines many short lines and curves into a network of splines.

If there are no additional notes necessary you may remove this entire section.
-->

* <!-- remove if no additional notes needed -->

[imperative mood]: https://chris.beams.io/posts/git-commit/#imperative

evanchaoli committed:
Enhance locks

Signed-off-by: Evan <chaol@vmware.com>

clarafu commented:
Hmm what kind of component enhancement were you thinking about adding? Because if this does cause one ATC to be overloaded then we might need to think about balancing the workloads out before getting this lock change merged

evanchaoli commented:
> Hmm what kind of component enhancement were you thinking about adding? Because if this does cause one ATC to be overloaded then we might need to think about balancing the workloads out before getting this lock change merged

@clarafu  First of all, this PR is very important. You know that our prod was unstable in Feb. We tried many ways to scale out: tuned container placement strategy, added 1 more ATC as well as some more workers, added more cpu and memory to the host where the db ran on. However the instability still lasted for a few months until I applied a private patch that gave VolumeCreationLock a separate mutex.

I found one ATC overloaded initially with a change that gave each lock type a separate mutex. But in this PR, I only give some lock types dedicate mutexes. I have been running this PR in a load test for a few months, it works well. Workloads can be distributed across ATC.

xtremerui commented on atc/db/lock/lock.go:
```diff
@@ -24,8 +24,39 @@ const (
+)
+
+func mapLockTypeToFactory(lockType int) int {
+	switch lockType {
```
Based on what it decides one lock type to have its own factory e.g. locks count, lock held duration ...?

evanchaoli replied:
Based on metrics. I just give those locks that are acquired big amount of times dedicate mutexes.

xtremerui commented on atc/atccmd/command.go:
```diff
@@ -1646,17 +1652,21 @@ type Closer interface {
-	}
+func constructLockConns(driverName, connectionString string) ([lock.FactoryCount]*sql.DB, error) {
+	conns := [lock.FactoryCount]*sql.DB{}
+	for i := 0; i < lock.FactoryCount; i++ {
```
Where is `FactoryCount` got initialized? What is an example value?

xtremerui commented on atc/db/lock/lock.go:
```diff
@@ -81,23 +116,29 @@ type lockFactory struct {
 
 func NewLockFactory(
-	conn *sql.DB,
+	conn [FactoryCount]*sql.DB,
```
```suggestion
	conns [FactoryCount]*sql.DB,
```

evanchaoli committed:
address review comment

Signed-off-by: Evan <chaol@vmware.com>
